Added queue position and approx wait time. #402

SuhanKoh opened the pull request.

Shredder121 commented:
IIRC embeds didn't look as nice.
Or was it decided to do that?
(Might have missed it)

Kurozume commented:
We decided against embeds because they don't look as nice as the current queue

SuhanKoh commented:
I have screenshots here, #402

Shredder121 commented:
Ok I see napster said it's okay

SuhanKoh commented:
Is there another way to display long message without using embeds? The issue here is once I added queue position and wait timer, the message gets too long :/

Shredder121 commented:
Without embeds there is more potential screen real estate (on desktop clients*)
embeds only grow as big as they do, so yeah..

napstr commented:
Embeds use a smaller font.

napstr commented:
Hi, can you provide screenshots please?

SuhanKoh commented:
<img width="581" alt="screen shot 2017-12-17 at 9 24 48 pm" src="https://user-images.githubusercontent.com/7716834/34091212-09b8c9ae-e371-11e7-8af4-b94633fe7d67.png">
<img width="631" alt="screen shot 2017-12-17 at 9 25 15 pm" src="https://user-images.githubusercontent.com/7716834/34091213-09ce3e1a-e371-11e7-83c6-f31becb7861f.png">
<img width="613" alt="screen shot 2017-12-17 at 9 25 40 pm" src="https://user-images.githubusercontent.com/7716834/34091214-09e4f664-e371-11e7-9630-6b29ffe49968.png">

SuhanKoh commented:
Squeezed to smallest possible on a macosx.
<img width="372" alt="screen shot 2017-12-17 at 9 30 13 pm" src="https://user-images.githubusercontent.com/7716834/34091293-8e662e4e-e371-11e7-999d-cef558078eff.png">

napstr commented:
> FredBoat shouldn't use embeds where it isn't useful. Not all users see them and it is inconsistent with our other messages.

I'd like to hear your counterproposal on displaying a lot of information without bloating the messages.

Frederikam commented:
>I'd like to hear your counterproposal on displaying a lot of information without bloating the messages.

With a normal message, possibly less verbose than the current one

SuhanKoh commented:
Should I concat the music name to maybe to x characters and replace it with "..." ?

SuhanKoh commented:
How does this look? 

<img width="1096" src="https://user-images.githubusercontent.com/7716834/34474853-f575d778-ef39-11e7-9b03-36c7d24ac247.png">

napstr commented:
@Frederikam :eyes:

Frederikam commented:
Instead of commas I would put linebreaks

napstr commented:
I think these seem to be equivalent but shorter replacements:
"has been seletected" -> "selected"
"position in queue" -> "position", or maybe just "queued"
"approximate wait time" -> "~ wait time", or maybe "playing in"

The last "minute" is completely wrong..can probably be replaced by single letter notation (h, m, s). We might have methods for that in TextUtils already.

SuhanKoh commented:
@Frederikam I think if I follow what napstr suggested, we don't need a line break. I will put up screenshots soon^tm.

SuhanKoh commented:
Any suggestion on how to put line break?

Maybe something like this?
```
Song selected #2 ~ Now playing
    HELLO! Songs Jukebox | Akhil Akkineni, Kalyani Priyadarshan | Vikram K Kumar | Anup Rubens (29:49) 

Song selected #1 ~ queued: #2, playing in: 29:49
    Adele - Hello (06:07)
```
<img width="895" alt="test" src="https://user-images.githubusercontent.com/7716834/34703469-f2e28a74-f4a7-11e7-98b1-507e015c5979.png">

Frederikam commented:
You're right, that actually looks better

I haven't reviewed the code, but I think Napster wanted to do that

napstr commented:
I like the song name on the second line too.

Nanabell commented:
I have a question. How are Livestreams handled here ?

SuhanKoh commented:
@Nanabell It will now only print once for livestream. Seems like the one in production will print twice. 
<img width="758" alt="livestream test" src="https://user-images.githubusercontent.com/7716834/34757785-58ce176e-f588-11e7-8aca-3ba63a2d5365.png">

SuhanKoh commented:
Screenshot fo the current change:
<img width="728" alt="two lines" src="https://user-images.githubusercontent.com/7716834/34757863-d8c8250e-f588-11e7-82c6-0b8d4ba70141.png">

Nanabell commented:
@SuhanKohI rather meant, if there is already a live stream in the queue/playing. What does the "playing in"  say when we have a live stream before it. since without a manual skip or the stream fucking up we would not play the songs after that

SuhanKoh commented:
@Nanabell I just checked, it will say Queued, but the time are not added from livestream since it's live. (let me check if there is a way to detect currently playing a live stream)

SuhanKoh commented:
<img width="758" alt="live stream check" src="https://user-images.githubusercontent.com/7716834/34805091-fa9f035c-f630-11e7-8202-2c4d75469cb0.png">

napstr commented on FredBoat/src/main/resources/lang/en_US.properties:
```diff
@@ -14,7 +14,11 @@ pauseSuccess=The player is now paused. You can unpause it with `{0}unpause`.
+selectTitleAndLength=**{0}** ({1})
+selectSuccessPartNowPlaying=*Now playing*
+selectSuccessPartQueueWaitTime=Queued: **{0}**, playing in: **{1}**.
+selectSuccessPartQueueHasStream=Queued: **{0}**, playing after live stream.
```
Please move all formatting from the translation files to the code.
https://github.com/Frederikam/FredBoat/blob/dev/CONTRIBUTING.md#rules-for-translations

SuhanKoh replied:
Oops, will do.

napstr commented on FredBoat/src/main/resources/lang/en_US.properties:
```diff
@@ -14,11 +14,11 @@ pauseSuccess=The player is now paused. You can unpause it with `{0}unpause`.
-selectSuccessPartQueueWaitTime=Queued: **{0}**, playing in: **{1}**.
-selectSuccessPartQueueHasStream=Queued: **{0}**, playing after live stream.
+selectSuccess=Song selected {0} {1}
+selectTitleAndLength={0} {1}
```
This is not translatable.

SuhanKoh replied:
Will fix in the next commit 👍

napstr commented on FredBoat/src/main/resources/lang/en_US.properties:
```diff
@@ -14,11 +14,11 @@ pauseSuccess=The player is now paused. You can unpause it with `{0}unpause`.
-selectSuccessPartNowPlaying=*Now playing*
-selectSuccessPartQueueWaitTime=Queued: **{0}**, playing in: **{1}**.
-selectSuccessPartQueueHasStream=Queued: **{0}**, playing after live stream.
+selectSuccess=Song selected {0} {1}
```
Why not merge these into a single {0} ? Or leave it off completely? Since the content of those variables does not look to be related to the String itself.

SuhanKoh replied:
I will merge this into one, "song selected" worth to be translated imo. I overlooked at the translation file, will fix.

napstr commented:
Two things:
- The "playing after livestream" message doesn't happen, when the livestream is in the queue, but not currently playing.
![image](https://user-images.githubusercontent.com/6048348/35638004-31ec62b0-06b6-11e8-955d-cebbee0f9068.png)

- Adding a track via link does not show any of the things done by this PR, but it should. 
![image](https://user-images.githubusercontent.com/6048348/35638093-7c109ac8-06b6-11e8-9398-c413e2b35f3b.png)

napstr commented:
More questions which you should test: What happens when a playlist/song is added
- as a youtube playlist
- as a spotify playlist
- as a hastebin link
- via ;;split

SuhanKoh commented:
Check list
* Fix queue message when there is a live stream in the queue.
* Fix queue message when playing via a direct youtube url.
* Verify youtube/spotify playlist also works
* Verify hastebin also works
* Verify ;;split command also works

SuhanKoh commented:
@napstr 
Doing youtube playlist, add hastebin link, and split command will retain their original message. Did you want me to change to performing the queue timer?

napstr commented:
I don't understand the question.

SuhanKoh commented:
What I mean is, current command of youtube playlist, hastebin list, and split command are showing their original message. Refer to your previous comment, did you want me to change those as well?

napstr commented:
Is this waiting on further review? Or is the code still WIP since the last one?

SuhanKoh commented:
I have updated the code, sorry for the wait. 
In this commit, I have made some assumption, so please let me know if this is ok. 
* When playing from playlist, hastebin, and split. I've set a max 5 queue messages, then print the counter, which is same as before.
<img width="463" alt="test" src="https://user-images.githubusercontent.com/7716834/36070644-443672d4-0eb4-11e8-9b0c-3cd8f84a3654.png">

SuhanKoh commented:
Screenshot for split
<img width="593" alt="splittest" src="https://user-images.githubusercontent.com/7716834/36070795-89386da8-0eb7-11e8-99ab-18a939bbfa18.png">

napstr commented:
Let's reduce it to 3 songs and a clue that more have been added like a line of 
> ...
between the details and the total count at the bottom?

@Frederikam needs to okay this.

SuhanKoh commented:
@Frederikam @napstr Let me know if this is appropriate. I added it because in the production, it usually reply to playlist with "Added 49 tracks. Found too many tracks to display." due to the > 800 characters check.

<img width="656" alt="splittest" src="https://user-images.githubusercontent.com/7716834/36132109-3008ae18-102a-11e8-9d2a-28aeb7ab8144.png">

Frederikam commented:
Sorry for the very late response. I haven't followed the conversation, so can you give me a summary of what changed? Is this the only thing that was changed? 
![image](https://user-images.githubusercontent.com/2582617/36661672-3b7d9f30-1adc-11e8-9c13-74b87cb1c626.png)

Shredder121 commented:
For `trackLoaded` it's also using the new format, so maybe also make a screenshot with loading a single track?
Then we can decide if it's what we would distribute to the masses.
Since I agree having a compact list of tracks, but I'd like to be sure the format if right for single tracks as well.

Does it also use the new format with searches? Just wondering.
(Also sorry for hijacking the review Fred 👀 you can continue)

SuhanKoh commented:
@Shredder121 

I loaded the same song to show how it looks like loading single track with empty queue and non-empty queue.

<img width="682" alt="singleload" src="https://user-images.githubusercontent.com/7716834/36707452-3b4ce880-1b23-11e8-9ed0-260a6c78e164.png">

SuhanKoh commented:
@Frederikam What was changed last time you reviewed was performing the commands below will also use the new printout, with queue and x time to play.

* as a youtube playlist
* as a spotify playlist
* as a hastebin link
* via ;;split

SuhanKoh commented:
@Frederikam I think I have made all your request changes.

Frederikam commented:
Unfortunately I don't feel like I have time to review this anytime soon, but at any rate, you should probably resolve those merge conflicts.

napstr commented:
Whats the point in solving merge conflicts over and over when the bottleneck is that there are no clear guidelines to the UI/UX of this?

Frederikam commented:
@napstr I don't believe I have as strong opinions about UX as you think I do. If you want you can review and merge this. As mentioned I haven't really followed the conversation.

Frederikam commented on FredBoat/src/main/resources/lang/en_US.properties:
```diff
@@ -14,7 +14,10 @@ pauseSuccess=The player is now paused. You can unpause it with `{0}unpause`.
-selectSuccess=Song **\#{0}** has been selected\: **{1}** ({2})
+selectSuccess=Song selected {0}
+selectSuccessPartNowPlaying=Now playing
+selectSuccessPartQueueWaitTime=Queued: {0}, playing in: {1}.
```
Though it may make more sense to say "Queued as #{0}"

napstr commented:
I don't want to review or manage anything UI/UX related.

SuhanKoh commented:
I don't think there are interest in merging this pr, so I will go ahead and close it

Frederikam commented:
@SuhanKoh the problem is that the sentinel branch is blocking this PR. Much of this code would probably need to be changed, and I might do that migration once we deploy the next major version of FredBoat.

This PR may not be merged, but I am keeping it open just so I won't forget.
